Filter API Docs (#9202)
hashicorp/consul: website/pages/api-docs/features/filtering.mdx

# Filtering

Filter expressions refine data queries for some API listing endpoints, as notated in the individual API documentation.

> To create a filter expression, you will write one or more expressions. Each expression has a matching operators composed of selectors and values.

Reviewer: The matching operators are the `==`, `!=`, `is`, etc. They don't contain the selectors and values themselves.

```suggestion
To create a filter expression, you will write one or more expressions. Each expression has a matching operators composed with selectors and values.
```

```shell

> curl -G <path> --data-urlencode 'filter=<"filter expression">'

Reviewer 2: ```suggestion
curl --get <path> --data-urlencode 'filter="<filter expression>"'
```
I feel the long form config option `--get` is a bit clearer than `-G`.

I realize this was already written like this prior to your change. I figure it might be worth changing while you're proposing these other modifications.

```

> Filtering is executed on the Consul server, before data is returned, reducing the network load. To pass a
> filter expression to Consul, with a data query, use the `filter` parameter.

Reviewer: Maybe the `with a data query` could be made more explicit:

```suggestion
Filtering is executed on the Consul server, before data is returned, reducing the network load. To pass a
filter expression to Consul, use the `filter` query parameter when sending requests to HTTP API endpoints that support it.
```

## Creating Expressions

A single expression is a matching operator with a selector and value. They are written in plain text format, boolean logic and parenthesization are
supported. In general whitespace is ignored, except within literal
strings.

### Matching Operators

All matching operators use a selector or value to choose what data should be
matched. Each endpoint that supports filtering accepts a potentially
different list of selectors and is detailed in the API documentation for
those endpoints.

```text
// Equality & Inequality checks
<Selector> == "<Value>"
<Selector> != "<Value>"

// Emptiness checks
<Selector> is empty
<Selector> is not empty

// Contains checks or Substring Matching

> <"Value"> in <Selector>
> <"Value"> not in <Selector>

Reviewer 2: ```suggestion
"<Value>" in <Selector>
"<Value>" not in <Selector>
```

### Values

Values are used by matching operators to create an expression. Values can be any valid selector, a number, or a string. It is best practice to quote values. 

Numbers can be any
base 10 integers and floating point numbers are possible. 

When quoting strings,
they may either be enclosed in double quotes or backticks. When enclosed in
backticks they are treated as raw strings and escape sequences such as `\n`
will not be expanded.

> ## Connect Expressions

Reviewer 2: ```suggestion
## Connecting Expressions
```
Should this be "connecting" so that its similar to the other headings?

Changing it might also remove any chance that someone misinterprets expressions as only being usable with Consul Connect.

Author: Agreed

**Command - Filtered**

```shell

> curl -G localhost:8500/v1/agent/services --data-urlencode 'filter=Meta.env == "qa"'

Reviewer 2: ```suggestion
curl --get localhost:8500/v1/agent/services --data-urlencode 'filter=Meta.env == "qa"'
```

**Command - Filtered**

```shell

> curl -G localhost:8500/v1/catalog/service/api-internal --data-urlencode 'filter=NodeMeta.os == "linux"'

Reviewer 2: ```suggestion
curl --get localhost:8500/v1/catalog/service/api-internal --data-urlencode 'filter=NodeMeta.os == "linux"'
```
